Userland/Libraries/LibBencode/Bencode.cpp

```diff
@@ -0,0 +1,176 @@
+/*
+ * Copyright (c) 2021, Cesar Torres <shortanemoia@protonmail.com>
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ */
+
+#include "Bencode.h"
+#include <AK/Format.h>
+#include <AK/Optional.h>
+#include <AK/OwnPtr.h>
+#include <AK/String.h>
+
+namespace Bencode {
+
+void BencodeEncoder::append(i64 value)
+{
+    m_builder.appendff("i{}e", value);
+}
+void BencodeEncoder::append(StringView const& value)
+{
+    m_builder.appendff("{}:{}", value.length(), value);
+}
+void BencodeEncoder::append(ReadonlyBytes const& value)
+{
+    m_builder.appendff("{}:", value.size());
+    for (auto byte : value)
+        m_builder.append(byte);
+}
+void BencodeEncoder::start_list()
+{
+    m_builder.append('l');
+}
+void BencodeEncoder::start_dict()
+{
+    m_builder.append('d');
+}
+void BencodeEncoder::end_list()
+{
+    m_builder.append('e');
+}
+void BencodeEncoder::end_dict()
+{
+    m_builder.append('e');
+}
+NonnullOwnPtr<ByteBuffer> BencodeEncoder::to_buffer()
+{
+    return make<ByteBuffer>(m_builder.to_byte_buffer());
+}
+void BencodeEncoder::clear()
+{
+    m_builder.clear();
+}
+
+OwnPtr<BencodeNode> decode_one(ReadonlyBytes const& buffer, size_t& index);
+OwnPtr<BencodeNode> decode_dict(ReadonlyBytes const& buffer, size_t& index);
+OwnPtr<BencodeNode> decode_list(ReadonlyBytes const& buffer, size_t& index);
+
+OwnPtr<BencodeNode> decode_list(ReadonlyBytes const& buffer, size_t& index)
+{
+    auto list = adopt_own_if_nonnull(new Vector<OwnPtr<BencodeNode>>());
+    while (buffer[index] != 'e') {
+
+        auto maybe_node = decode_one(buffer, index);
+        if (!maybe_node)
+            return {};
+        list->append(move(maybe_node));
+    }
+    index++;
+    return adopt_own_if_nonnull(new BencodeNode { BencodeNodeType::List, move(list) });
+}
+
+OwnPtr<BencodeNode> decode_dict(ReadonlyBytes const& buffer, size_t& index)
+{
+    auto dict = adopt_own_if_nonnull(new Vector<OwnPtr<BencodeNode>>());
+    while (buffer[index] != 'e') {
+        auto maybe_key = decode_one(buffer, index);
+        if (!maybe_key)
+            return {};
+        if (maybe_key->type != BencodeNodeType::String)
+            return {};
+        auto maybe_value = decode_one(buffer, index);
+        if (!maybe_value)
+            return {};
+        dict->append(adopt_own_if_nonnull(new BencodeNode { BencodeNodeType::KeyValuePair, adopt_own_if_nonnull(new KeyValuePair { move(maybe_key->data.get<String>()), move(maybe_value) }) }));
+    }
+    index++;
+    return adopt_own_if_nonnull(new BencodeNode { BencodeNodeType::Dictionary, move(dict) });
+}
+
+OwnPtr<BencodeNode> decode_one(ReadonlyBytes const& buffer, size_t& index)
+{
+
+    switch (buffer[index++]) {
+    case 'i': {
+        char* end;
+        if (buffer[index] == '0')
+            return {};
+        i64 value = strtol(reinterpret_cast<const char*>(buffer.slice(index).data()), &end, 10);
```

Comment:
Please use east const style as the project mandates, so `char const*` instead of `const char*`.

Reply:
Sidenote: Since this is base 10, a more appropriate way to parse it would be `StringView(buffer.slice(index)).to_int<i64>()`.
unless you're using `end`, in which case, fiiiine

Reply:
I'm using end :^)